File: src/main/java/hzt/aoc/day20/Day20Challenge.java
package hzt.aoc.day20;

import hzt.aoc.Challenge;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public abstract class Day20Challenge extends Challenge {

    Day20Challenge(String challengeTitle, String description) {
        super(challengeTitle, description, "20201220-input-day20ref.txt");
    }

    @Override
    protected String solve(List<String> inputList) {
        final Map<Integer, List<List<Boolean>>> tileIdsToGrids = new HashMap<>();
        List<List<Boolean>> grid = new ArrayList<>();
        int tileId = 0;
        for (String line : inputList) {
            if (line.contains("Tile")) {
                tileId = Integer.parseInt(line.replace("Tile ", "").replace(":", "").strip());
                grid = new ArrayList<>();
            } else if (!line.isEmpty()) {
                List<Boolean> row = line.chars().mapToObj(c -> (char) c).map(c -> c == '#').collect(Collectors.toList());
                grid.add(row);
            }
            if (line.isEmpty()) {
                tileIdsToGrids.put(tileId, grid);
            }
        }
        tileIdsToGrids.forEach((k, v) -> LOGGER.info(k + "->" + booleanGrid2DAsString(v)));
        return getMessage(calculateAnswer(tileIdsToGrids));
    }

    protected abstract long calculateAnswer(Map<Integer, List<List<Boolean>>> tileIdsToGrids);


    abstract String getMessage(long value);
}

File: src/main/kotlin/hzt/aoc/day20/Day20Challenge.kt
package hzt.aoc.day20

import hzt.aoc.Challenge

abstract class Day20Challenge internal constructor(challengeTitle: String, description: String) :
    Challenge(challengeTitle, description, "20201220-input-day20.txt") {
    override fun solve(inputList: List<String>): String {
        val tileIdsToGrids = parseInput(inputList)
        return getMessage(calculateAnswer(tileIdsToGrids))
    }

    private fun parseInput(inputList: List<String>): Map<Int, Tile> {
        val tileIdsToGrids: MutableMap<Int, Tile> = HashMap()
        var tileContent: MutableList<String> = ArrayList()
        var tileId = 0
        for (line in inputList) {
            if (line.contains("Tile")) {
                tileId = line.replace("Tile ", "").replace(":", "").trim().toInt()
                tileContent = ArrayList()
            } else if (line.isNotEmpty()) {
                tileContent.add(line)
            }
            if (line.isBlank()) {
                tileIdsToGrids[tileId] = Tile(tileContent)
            }
        }
        tileIdsToGrids.forEach { (id, tile) -> LOGGER.trace("$id->$tile") }
        LOGGER.trace(tileIdsToGrids.size)
        return tileIdsToGrids
    }

    protected abstract fun calculateAnswer(tileIdsToGrids: Map<Int, Tile>): Long
    abstract fun getMessage(value: Long): String
}
